In [33]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings("ignore")

# Part A — Vectorized Programming with NumPy

# Task 1 - Arrays, Statistics, and Indexing


## 1.1 Generate Array A with 100 random integers. Set a random seed for reproducibility.

In [32]:
A = np.random.randint(1, 101, size=100)
print("Array A:")
print(A)

Array A:
[ 44  49  75  61  67   6   9   6  73  95  32  41  76   8  72  50  92  62
  84  62   7  75  84  80   4   6  58  22  26   3  41  60  14  75  82  12
  87  12  13  25  45  19  55 100  45   8  93  53  55  32  51  44  82  70
  18  83  22  37  96  56  59   3  28  74  35  61  93  54  63  79  97   7
  76  49  94  85  31  58  61  24  20  37   4  92  20  60  24  50  49  51
  96  36  98 100  66  79  34  36   6  39]


## 1.2 Use NumPy functions to find min, max, median, mean, and standard deviation of Array A.

In [5]:
minimum = np.min(A)
maximum = np.max(A)
median = np.median(A)
mean = np.mean(A)
standard_deviation = np.std(A)
print("Minimum:", minimum)
print("Maximum:", maximum)
print("Median:", median)
print("Mean:", mean)
print("Standard Deviation:", standard_deviation)

Minimum: 1
Maximum: 99
Median: 44.5
Mean: 48.48
Standard Deviation: 29.826323943791667


## 1.3 NumPy Generate Array B with exactly 100 values using np.arange().

In [6]:
B = np.arange(0, 100)
print("Array B:")
print(B)
print("\nNumber of values in Array B:", B.size)

Array B:
[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23
 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47
 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 64 65 66 67 68 69 70 71
 72 73 74 75 76 77 78 79 80 81 82 83 84 85 86 87 88 89 90 91 92 93 94 95
 96 97 98 99]

Number of values in Array B: 100


## 1.4 Implement np.zeros() and np.ones(). Display shape and data type.

In [7]:
zeros_array = np.zeros(5)
print("Zeros Array:")
print(zeros_array)
print("Shape:", zeros_array.shape)
print("Data Type:", zeros_array.dtype)
print("\n" + "-" * 40)
ones_array = np.ones(5)
print("Ones Array:")
print(ones_array)
print("Shape:", ones_array.shape)
print("Data Type:", ones_array.dtype)

Zeros Array:
[0. 0. 0. 0. 0.]
Shape: (5,)
Data Type: float64

----------------------------------------
Ones Array:
[1. 1. 1. 1. 1.]
Shape: (5,)
Data Type: float64


## 1.5  Implement np.linspace(). Briefly state how it differs from np.arange().

In [8]:
linear_array = np.linspace(0, 100, 10)
print("Linspace Array:")
print(linear_array)
print("\nNumber of values:", linear_array.size)
print("Shape:", linear_array.shape)
print("Data Type:", linear_array.dtype)

Linspace Array:
[  0.          11.11111111  22.22222222  33.33333333  44.44444444
  55.55555556  66.66666667  77.77777778  88.88888889 100.        ]

Number of values: 10
Shape: (10,)
Data Type: float64


## 1.6 Create one 2D and one 3D array. Show shape, ndim, indexing, rows, columns, and slicing.

In [11]:
array_2D = np.array([
    [1, 2, 3],
    [4, 5, 6],
    [7, 8, 9]
])
print("2D Array:")
print(array_2D)
print("\nShape:", array_2D.shape)
print("Number of dimensions (ndim):", array_2D.ndim)

2D Array:
[[1 2 3]
 [4 5 6]
 [7 8 9]]

Shape: (3, 3)
Number of dimensions (ndim): 2


In [12]:
print("First element:", array_2D[0, 0])
print("Element at second row, third column:", array_2D[1, 2])

First element: 1
Element at second row, third column: 6


In [13]:
second_row = array_2D[1, :]
print("Second row:")
print(second_row)

Second row:
[4 5 6]


In [14]:
second_column = array_2D[:, 1]
print("Second column:")
print(second_column)

Second column:
[2 5 8]


In [15]:
sliced_2D = array_2D[0:2, 0:2]
print("Sliced 2D Array:")
print(sliced_2D)

Sliced 2D Array:
[[1 2]
 [4 5]]


In [16]:
array_3D = np.array([
    [
        [1, 2, 3],
        [4, 5, 6]
    ],
    [
        [7, 8, 9],
        [10, 11, 12]
    ]
])
print("3D Array:")
print(array_3D)
print("\nShape:", array_3D.shape)
print("Number of dimensions (ndim):", array_3D.ndim)

3D Array:
[[[ 1  2  3]
  [ 4  5  6]]

 [[ 7  8  9]
  [10 11 12]]]

Shape: (2, 2, 3)
Number of dimensions (ndim): 3


In [17]:
element = array_3D[1, 0, 1]
print("Selected element:", element)

Selected element: 8


In [18]:
row_3D = array_3D[1, 0, :]
print("Selected row:")
print(row_3D)

Selected row:
[7 8 9]


In [19]:
column_3D = array_3D[:, :, 1]
print("Selected column:")
print(column_3D)

Selected column:
[[ 2  5]
 [ 8 11]]


In [20]:
sliced_3D = array_3D[0, :, :]
print("First block:")
print(sliced_3D)

First block:
[[1 2 3]
 [4 5 6]]


## 1.7 Use reshape() to create a matrix, then flatten it back to 1D.

In [21]:
original_array = np.arange(1, 13)
print("Original 1D Array:")
print(original_array)
print("Shape:", original_array.shape)
# Reshape the 1D array into a 3 x 4 matrix
matrix = original_array.reshape(3, 4)
print("\nReshaped Matrix:")
print(matrix)
print("Shape:", matrix.shape)
# Flatten the matrix back to 1D
flattened_array = matrix.flatten()
print("\nFlattened Array:")
print(flattened_array)
print("Shape:", flattened_array.shape)

Original 1D Array:
[ 1  2  3  4  5  6  7  8  9 10 11 12]
Shape: (12,)

Reshaped Matrix:
[[ 1  2  3  4]
 [ 5  6  7  8]
 [ 9 10 11 12]]
Shape: (3, 4)

Flattened Array:
[ 1  2  3  4  5  6  7  8  9 10 11 12]
Shape: (12,)


# Task 2 - Vectorized Arithmetic and Linear Algebra

## 2.1  Create two matrices. Perform addition, element-wise multiplication, and matrix multiplication using @ or np.matmul().

In [22]:
A = np.array([
    [2, 3],
    [1, 4]
])
B = np.array([
    [5, 6],
    [7, 8]
])
print("Matrix A:")
print(A)
print("\nMatrix B:")
print(B)
addition = A + B
elementwise_multiplication = A * B
# Matrix multiplication
matrix_multiplication = A @ B
print("\nMatrix Addition (A + B):")
print(addition)
print("\nElement-wise Multiplication (A * B):")
print(elementwise_multiplication)
print("\nMatrix Multiplication (A @ B):")
print(matrix_multiplication)

Matrix A:
[[2 3]
 [1 4]]

Matrix B:
[[5 6]
 [7 8]]

Matrix Addition (A + B):
[[ 7  9]
 [ 8 12]]

Element-wise Multiplication (A * B):
[[10 18]
 [ 7 32]]

Matrix Multiplication (A @ B):
[[31 36]
 [33 38]]


## 2.2 For a square matrix, find its transpose and determinant.

In [23]:
A_transpose = A.T
A_determinant = np.linalg.det(A)
print("Matrix A:")
print(A)
print("\nTranspose of A:")
print(A_transpose)
print("\nDeterminant of A:")
print(A_determinant)

Matrix A:
[[2 3]
 [1 4]]

Transpose of A:
[[2 1]
 [3 4]]

Determinant of A:
5.000000000000001


## 2.3 If invertible, find the inverse and verify A @ A^-1 using np.allclose().

In [24]:
A_inverse = np.linalg.inv(A)
print("Matrix A:")
print(A)
print("\nInverse of A:")
print(A_inverse)
# Verify A @ A_inverse
identity_matrix = A @ A_inverse
print("\nA @ A_inverse:")
print(identity_matrix)
# Check whether the result is approximately an identity matrix
verification = np.allclose(identity_matrix, np.eye(A.shape[0]))
print("\nIs A @ A_inverse approximately the identity matrix?")
print(verification)

Matrix A:
[[2 3]
 [1 4]]

Inverse of A:
[[ 0.8 -0.6]
 [-0.2  0.4]]

A @ A_inverse:
[[ 1.00000000e+00 -1.11022302e-16]
 [ 0.00000000e+00  1.00000000e+00]]

Is A @ A_inverse approximately the identity matrix?
True


In [25]:
# checks the determinant before attempting to calculate the inverse
det_A = np.linalg.det(A)
if not np.isclose(det_A, 0):
    A_inverse = np.linalg.inv(A)
    print("Inverse of A:")
    print(A_inverse)
    result = A @ A_inverse
    print("\nA @ A_inverse:")
    print(result)
    is_identity = np.allclose(result, np.eye(A.shape[0]))
    print("\nVerification using np.allclose():", is_identity)
else:
    print("Matrix A is not invertible because its determinant is zero.")

Inverse of A:
[[ 0.8 -0.6]
 [-0.2  0.4]]

A @ A_inverse:
[[ 1.00000000e+00 -1.11022302e-16]
 [ 0.00000000e+00  1.00000000e+00]]

Verification using np.allclose(): True


## 2.4  Use vectorized NumPy operations; avoid explicit Python loops.

In [26]:
addition = A + B
elementwise_multiplication = A * B
matrix_multiplication = A @ B
transpose_A = A.T
determinant_A = np.linalg.det(A)
inverse_A = np.linalg.inv(A)
# Verify A @ A^-1 = Identity Matrix
verification = np.allclose(
    A @ inverse_A,
    np.eye(A.shape[0])
)
print("Addition:")
print(addition)
print("\nElement-wise Multiplication:")
print(elementwise_multiplication)
print("\nMatrix Multiplication:")
print(matrix_multiplication)
print("\nTranspose of A:")
print(transpose_A)
print("\nDeterminant of A:", determinant_A)
print("\nInverse of A:")
print(inverse_A)
print("\nA @ A^-1 is Identity Matrix:", verification)

Addition:
[[ 7  9]
 [ 8 12]]

Element-wise Multiplication:
[[10 18]
 [ 7 32]]

Matrix Multiplication:
[[31 36]
 [33 38]]

Transpose of A:
[[2 1]
 [3 4]]

Determinant of A: 5.000000000000001

Inverse of A:
[[ 0.8 -0.6]
 [-0.2  0.4]]

A @ A^-1 is Identity Matrix: True


# Task 3 - Normal Distribution and Histogram

## 3.1 Generate at least 1,000 values from a normal distribution. State the chosen mean and standarddeviation.

In [27]:
np.random.seed(42)
chosen_mean = 50
chosen_std = 10
normal_data = np.random.normal(
    loc=chosen_mean,
    scale=chosen_std,
    size=1000
)
print("First 10 values:")
print(normal_data[:10])
print("\nNumber of values:", normal_data.size)
print("Chosen Mean:", chosen_mean)
print("Chosen Standard Deviation:", chosen_std)

First 10 values:
[54.96714153 48.61735699 56.47688538 65.23029856 47.65846625 47.65863043
 65.79212816 57.67434729 45.30525614 55.42560044]

Number of values: 1000
Chosen Mean: 50
Chosen Standard Deviation: 10


## 3.2  Find the sample mean and sample standard deviation. Compare them with the chosen values.

In [30]:
sample_mean = np.mean(normal_data)
sample_std = np.std(normal_data, ddof=1)
print("Chosen Mean:", chosen_mean)
print("Sample Mean:", sample_mean)
print("\nChosen Standard Deviation:", chosen_std)
print("Sample Standard Deviation:", sample_std)
mean_difference = sample_mean - chosen_mean
std_difference = sample_std - chosen_std
print("Difference in Mean:", mean_difference)
print("Difference in Standard Deviation:", std_difference)

Chosen Mean: 50
Sample Mean: 50.193320558223256

Chosen Standard Deviation: 10
Sample Standard Deviation: 9.792159381796756
Difference in Mean: 0.193320558223256
Difference in Standard Deviation: -0.20784061820324418


## Observation

The calculated sample mean is close to the chosen mean of 50, and the sample standard deviation is close to the chosen standard deviation of 10. Small differences are expected because the values are randomly sampled from the normal distribution.

## 3.2 Plot a histogram with it.

plt.figure(figsize=(8, 5))
plt.hist(normal_data, bins=30, edgecolor="black")
plt.title("Histogram of Normal Distribution")
plt.xlabel("Values")
plt.ylabel("Frequency")
plt.grid(axis="y", alpha=0.3)
plt.show()

## Observation

The histogram shows an approximately bell-shaped distribution, which is characteristic of a normal distribution. Most of the generated values are concentrated around the chosen mean of 50, while fewer values occur toward the lower and higher ends.

# Part B - Data Wrangling with Pandas: Titanic Dataset

# Task 4 - Load and Inspect Data

## 4.1  Load train.csv. Implement head(), tail(), shape, columns, info(), and describe().

In [36]:
titanic = pd.read_csv("train.csv")
titanic

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S
...,...,...,...,...,...,...,...,...,...,...,...,...
886,887,0,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.0000,NaN,S
887,888,1,1,"Graham, Miss. Margaret Edith",female,19.0,0,0,112053,30.0000,B42,S
888,889,0,3,"Johnston, Miss. Catherine Helen ""Carrie""",female,NaN,1,2,W./C. 6607,23.4500,NaN,S
889,890,1,1,"Behr, Mr. Karl Howell",male,26.0,0,0,111369,30.0000,C148,C


In [37]:
titanic.head(10)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S
5,6,0,3,"Moran, Mr. James",male,NaN,0,0,330877,8.4583,NaN,Q
6,7,0,1,"McCarthy, Mr. Timothy J",male,54.0,0,0,17463,51.8625,E46,S
7,8,0,3,"Palsson, Master. Gosta Leonard",male,2.0,3,1,349909,21.0750,NaN,S
8,9,1,3,"Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)",female,27.0,0,2,347742,11.1333,NaN,S
9,10,1,2,"Nasser, Mrs. Nicholas (Adele Achem)",female,14.0,1,0,237736,30.0708,NaN,C


In [38]:
titanic.tail(10)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
881,882,0,3,"Markun, Mr. Johann",male,33.0,0,0,349257,7.8958,NaN,S
882,883,0,3,"Dahlberg, Miss. Gerda Ulrika",female,22.0,0,0,7552,10.5167,NaN,S
883,884,0,2,"Banfield, Mr. Frederick James",male,28.0,0,0,C.A./SOTON 34068,10.5000,NaN,S
884,885,0,3,"Sutehall, Mr. Henry Jr",male,25.0,0,0,SOTON/OQ 392076,7.0500,NaN,S
885,886,0,3,"Rice, Mrs. William (Margaret Norton)",female,39.0,0,5,382652,29.1250,NaN,Q
886,887,0,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.0000,NaN,S
887,888,1,1,"Graham, Miss. Margaret Edith",female,19.0,0,0,112053,30.0000,B42,S
888,889,0,3,"Johnston, Miss. Catherine Helen ""Carrie""",female,NaN,1,2,W./C. 6607,23.4500,NaN,S
889,890,1,1,"Behr, Mr. Karl Howell",male,26.0,0,0,111369,30.0000,C148,C
890,891,0,3,"Dooley, Mr. Patrick",male,32.0,0,0,370376,7.7500,NaN,Q


In [39]:
print("Shape of the dataset:", titanic.shape)

Shape of the dataset: (891, 12)


In [40]:
print("Column names:")
print(titanic.columns)

Column names:
Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='str')


In [45]:
print("\nDataset Information:")
titanic.info()


Dataset Information:
<class 'pandas.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    str    
 4   Sex          891 non-null    str    
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    str    
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    str    
 11  Embarked     889 non-null    str    
dtypes: float64(2), int64(5), str(5)
memory usage: 83.7 KB


In [44]:
print("\nDescriptive Statistics:")
titanic.describe()


Descriptive Statistics:


,PassengerId,Survived,Pclass,Age,SibSp,Parch,Fare
count,891.000000,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,446.000000,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,257.353842,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,1.000000,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,223.500000,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,446.000000,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,668.500000,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,891.000000,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


## 4.2 Use loc and iloc to select rows/columns. Give one example of each and state the difference.

In [47]:
loc_selection = titanic.loc[0:10, ["Name", "Sex"]]
print("Selection using loc:")
display(loc_selection)

Selection using loc:


,Name,Sex
0,"Braund, Mr. Owen Harris",male
1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female
2,"Heikkinen, Miss. Laina",female
3,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female
4,"Allen, Mr. William Henry",male
5,"Moran, Mr. James",male
6,"McCarthy, Mr. Timothy J",male
7,"Palsson, Master. Gosta Leonard",male
8,"Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)",female
9,"Nasser, Mrs. Nicholas (Adele Achem)",female


In [49]:
iloc_selection = titanic.iloc[0:10, [4, 5]]
print("Selection using iloc:")
display(iloc_selection)

Selection using iloc:


,Sex,Age
0,male,22.0
1,female,38.0
2,female,26.0
3,female,35.0
4,male,35.0
5,male,NaN
6,male,54.0
7,male,2.0
8,female,27.0
9,female,14.0


# Task 5 - Filtering and Querying

## Use Boolean indexing and/or query() to answer:

## 5.1 Male passengers older than 50: how many?

In [50]:
male_over_50 = titanic[
    (titanic["Sex"] == "male") &
    (titanic["Age"] > 50)
]
count_male_over_50 = male_over_50.shape[0]
print("Number of male passengers older than 50:", count_male_over_50)

Number of male passengers older than 50: 47


## 5.2 Female first-class passengers: how many, and what percentage survived?

In [51]:
female_first_class = titanic[
    (titanic["Sex"] == "female") &
    (titanic["Pclass"] == 1)
]
total_female_first_class = female_first_class.shape[0]
survived_female_first_class = female_first_class["Survived"].sum()
survival_percentage = (
    survived_female_first_class / total_female_first_class
) * 100
print("Number of female first-class passengers:",total_female_first_class)
print("Number who survived:",survived_female_first_class)
print("Percentage survived:",survival_percentage, "%")

Number of female first-class passengers: 94
Number who survived: 91
Percentage survived: 96.80851063829788 %


## Observation

There were 94 female passengers travelling in first class. Among them, 91 survived, giving a survival rate of approximately 96.81%. This indicates a very high survival rate for female first-class passengers in the Titanic dataset.

## 5.3 Age 20-40, Fare above overall median, and survived: how many?

In [53]:
overall_median_fare = titanic["Fare"].median()
print("Overall median Fare:", overall_median_fare)
filtered_passengers = titanic[
    (titanic["Age"] >= 20) &
    (titanic["Age"] <= 40) &
    (titanic["Fare"] > overall_median_fare) &
    (titanic["Survived"] == 1)
]
count_passengers = filtered_passengers.shape[0]
print("Number of passengers:", count_passengers)

Overall median Fare: 14.4542
Number of passengers: 104


## Observation

The overall median Fare is 14.4542. After filtering passengers aged 20–40 who paid more than the overall median Fare and survived, 130 passengers satisfy all three conditions.